src/encoding_transcode.rs

```diff
@@ -0,0 +1,432 @@
+use bytes::{Bytes, BytesMut};
+use encoding_rs::{CoderResult, Encoding};
+
+const BUFFER_SIZE: usize = 4096;
+
+// BOM unicode character (U+FEFF) expressed in utf-8
+// http://unicode.org/faq/utf_bom.html#bom4
+const BOM_UTF8: &[u8] = b"\xef\xbb\xbf";
+const BOM_UTF8_LEN: usize = BOM_UTF8.len();
+
+/// Helps transcoding from the specified encoding to utf8
+pub struct Decoder {
+    buffer: [u8; BUFFER_SIZE],
+    output: BytesMut,
+    inner: encoding_rs::Decoder,
+}
+
+impl Decoder {
+    pub fn new(encoding: &'static Encoding) -> Self {
+        Self {
+            buffer: [0; BUFFER_SIZE],
+            output: BytesMut::new(),
+            // We explicitly choose not to remove BOM as part of encoding_rs's
+            // decoding capabilities: the library has support for it, but it does
+            // so only for the first input provided to the decoder (basically,
+            // start of the stream), and for our usecases, we may get BOM markers
+            // in later inputs too (eg: when reading multiple files):
+            // https://docs.rs/encoding_rs/0.8.26/encoding_rs/struct.Encoding.html#method.new_decoder_with_bom_removal
+            //
+            // We can try to maintain separate decoders for each unique stream
+            // (eg: by filepath when reading multiple files), but that mandates
+            // cleanup of the initialized decoder structs/buffers when they are
+            // no longer needed (eg: when files are closed), which can get
+            // complicated. So we opt for simplicity here.
+            //
+            // BOM markers are still removed if the input starts with it:
+            // see decode_to_utf8() for the rationale/logic.
+            inner: encoding.new_decoder_without_bom_handling(),
+        }
+    }
+
+    pub fn decode_to_utf8(&mut self, input: Bytes) -> Bytes {
+        let mut total_read_from_input = 0;
+        let mut total_had_errors = false;
+
+        loop {
+            let (result, read, written, had_errors) = self.inner.decode_to_utf8(
+                &input[total_read_from_input..],
+                &mut self.buffer,
+                false, // not last (since we are processing a continous stream)
+            );
+
+            total_read_from_input += read;
+            total_had_errors |= had_errors;
+
+            self.output.extend_from_slice(&self.buffer[..written]);
+
+            match result {
+                CoderResult::InputEmpty => {
+                    // we have consumed all of the given input so we are done!
+                    break;
+                }
+                CoderResult::OutputFull => {
+                    continue;
+                }
```

Comment:
The exhaustive match is good. However, this is the end of the loop, so the `continue` appears superfluous:
```suggestion
                CoderResult::OutputFull => ()
```